eras/alonzo/impl/testlib/Test/Cardano/Ledger/Alonzo/Imp/UtxowSpec/Invalid.hs

```diff
@@ -185,42 +214,136 @@ spec = describe "Invalid transactions" $ do
+          , injectFailure $
+              CollectErrors [NoRedeemer $ mkSpendingPurpose $ AsItem txIn]
+          ]
 
   it "Phase 1 script failure" $ do
-    const $ pendingWith "not implemented yet"
+    -- Script will be invalid because slot 100 will be in the future
+    scriptHash <- impAddNativeScript $ mkTimeStart 100
+    txIn <- produceScript scriptHash
+    let tx = mkBasicTx mkBasicTxBody & bodyTxL . inputsTxBodyL .~ [txIn]
+    submitFailingTx tx [injectFailure $ ScriptWitnessNotValidatingUTXOW [scriptHash]]
 
   it "Valid transaction marked as invalid" $ do
-    const $ pendingWith "not implemented yet"
+    let tx = mkBasicTx mkBasicTxBody & isValidTxL .~ IsValid False
+    submitFailingTx tx [injectFailure (ValidationTagMismatch (IsValid False) PassedUnexpectedly)]
 
   it "Invalid transaction marked as valid" $ do
-    const $ pendingWith "not implemented yet"
+    let scriptHash = withSLanguage PlutusV1 (hashPlutusScript . alwaysFailsWithDatum)
+    txIn <- produceScript scriptHash
+    let tx = mkBasicTx mkBasicTxBody & bodyTxL . inputsTxBodyL .~ [txIn]
+    expectPhase2Invalid tx
 
   it "Too many execution units for tx" $ do
-    const $ pendingWith "not implemented yet"
+    let scriptHash = withSLanguage PlutusV1 (hashPlutusScript . alwaysSucceedsWithDatum)
+    txIn <- produceScript scriptHash
+    maxExUnits <- getsNES $ nesEsL . curPParamsEpochStateL . ppMaxTxExUnitsL
+    let
+      txExUnits = maxExUnits <> ExUnits 1 1
+      prp = mkSpendingPurpose (AsIx 0)
+      dat = Data $ P.I 32
+      tx =
+        mkBasicTx mkBasicTxBody
+          & bodyTxL . inputsTxBodyL .~ [txIn]
+          & witsTxL . rdmrsTxWitsL . redeemersL <>~ Map.singleton prp (dat, txExUnits)
+    submitFailingTx tx [injectFailure $ ExUnitsTooBigUTxO maxExUnits txExUnits]
 
-  it "Missing signature for collateral input" $ do
-    const $ pendingWith "not implemented yet"
+  it "Missing witness for collateral input" $ do
+    let scriptHash = withSLanguage PlutusV1 (hashPlutusScript . alwaysSucceedsWithDatum)
+    scriptInput <- produceScript scriptHash
+    (collateralHash, collateralAddr) <- freshKeyAddr
+    collateralInput <- sendCoinTo collateralAddr $ Coin 1_000_000
+    let
+      tx =
+        mkBasicTx mkBasicTxBody
+          & bodyTxL . inputsTxBodyL <>~ [scriptInput]
+          & bodyTxL . collateralInputsTxBodyL <>~ [collateralInput]
+      isOtherWitness wit = witVKeyHash wit /= asWitness collateralHash
+      dropCollateralWitness = witsTxL . addrTxWitsL %~ Set.filter isOtherWitness
+    withPostFixup (pure . dropCollateralWitness) $
+      submitFailingTx
+        tx
+        [injectFailure $ MissingVKeyWitnessesUTXOW [asWitness collateralHash]]
 
   it "Insufficient collateral" $ do
-    const $ pendingWith "not implemented yet"
+    let scriptHash = withSLanguage PlutusV1 (hashPlutusScript . alwaysSucceedsWithDatum)
+    scriptInput <- produceScript scriptHash
+    let collateral = Coin 123
```

Comment:
Why not set this relative to the expected one, rather than hardcode it, to make the test more robust?

Reply:
The minimum required collateral is calculated from the fee, which we don't know until the tx is fixed up. I assume this would then require an iterative process until the values settle to a fixed point, similar to what we do in some of the fixup functions (eg for the fees themselves) and I think this would be more complicated than needed for this test. I think the value I've given (`123`) is always going to be a lot less than the minimum.

However, perhaps modifying the collateral in a post-fixup function wouldn't change the fee? If so, then I'd just have to add recalculation of the body hash using `fixupPPHash`.

BTW, the collateral created by the `addCollateralInput` fixup function is just set to a large, fixed number (`10_000_000`) and there's a `TODO: make more accurate` :-) I was wondering about improving that before finishing this PR, but I think it would have the same circularity problem.

Reply:
> I was wondering about improving that before finishing this PR, but I think it would have the same circularity problem.

Yes, this is a much more complex issue that it seems at first. It would deserve its own PR. Let's postpone this until later